File: src/host/acp/subagents/wire.ts
// Extension session updates through SDK 1.4.0: zSessionUpdate is a closed union, so an unknown `sessionUpdate`
// kind makes parseParams throw and the notification never reaches onUpdate. rewriteExtension swaps the variant
// for `session_info_update` (every field optional, `_meta` preserved) with the original update parked under
// EXT_META_KEY — keeping the same `session/update` handler means wire order against adjacent updates is preserved.

import type * as acp from '@agentclientprotocol/sdk';
import type { SubagentState } from '@shared/subagents';
import type { AsyncTaskState } from '@shared/transcript';

export const EXT_META_KEY = 'acpira/extension';

const EXT_KINDS = new Set([
  'subagent_update',          // RFD #1992 upsert
  'subagent_spawned',         // claude-agent-acp 0.78/0.79
  'subagent_state_update',    // claude-agent-acp 0.78/0.79
  'async_task_spawned',       // JetBrains AIR asyncTasks (claude-agent-acp acp-subagents, codex-acp background terminal)
  'async_task_progress',
  'async_task_state_update',
]);

export interface SubagentLifecycle {
  kind: 'lifecycle';
  peerSessionId: string;
  title?: string;
  task?: string;
  capabilities?: { cancel?: boolean };
  state?: SubagentState;
  meta?: Record<string, unknown>;
}

// One asyncTasks update decoded: the three kinds share asyncTaskId; spawned carries the card fields,
// progress the live ones, state the transition. toolCallId links the task to the tool row that spawned
// it and can show up on any of the three, including after the task was first announced
export interface AsyncTaskEvent {
  kind: 'async_task';
  event: 'spawned' | 'progress' | 'state';
  asyncTaskId: string;
  name?: string;
  taskType?: string;
  description?: string;
  showInTranscript?: boolean;
  canStop?: boolean;
  outputFilePath?: string;
  toolCallId?: string;
  summary?: string;
  lastToolName?: string;
  usage?: { totalTokens?: number; toolUses?: number; durationMs?: number };
  state?: AsyncTaskState;
  meta: Record<string, unknown>;
}

export type ExtensionUpdate = SubagentLifecycle | AsyncTaskEvent | { kind: 'ignored'; sessionUpdate: string };

const ASYNC_TASK_STATES: readonly AsyncTaskState[] = ['running', 'paused', 'completed', 'failed', 'stopped'];
const ASYNC_TASK_KINDS: Record<string, AsyncTaskEvent['event']> = {
  async_task_spawned: 'spawned',
  async_task_progress: 'progress',
  async_task_state_update: 'state',
};

function record(v: unknown): Record<string, unknown> | undefined {
  return v !== null && typeof v === 'object' && !Array.isArray(v) ? v as Record<string, unknown> : undefined;
}

function str(v: unknown): string | undefined {
  return typeof v === 'string' && v.length > 0 ? v : undefined;
}

function stateOf(v: unknown, log?: (line: string) => void): SubagentState | undefined {
  if (v === undefined || v === null) return undefined;
  if (v === 'running' || v === 'completed' || v === 'failed' || v === 'cancelled' || v === 'disconnected') return v;
  log?.(`unknown subagent state ${JSON.stringify(v)} — treating as running`);
  return 'running';
}

function capabilitiesOf(v: unknown): { cancel?: boolean } | undefined {
  const c = record(v);
  return c === undefined ? undefined : { cancel: c.cancel === true };
}

// Inbound `session/update` notification → rewritten for the closed SDK union; everything else passes through
export function rewriteExtension(msg: unknown): unknown {
  const m = record(msg);
  if (m === undefined || m.method !== 'session/update' || 'id' in m) return msg;
  const params = record(m.params);
  const update = record(params?.update);
  if (update === undefined || typeof update.sessionUpdate !== 'string' || !EXT_KINDS.has(update.sessionUpdate)) return msg;
  return { ...m, params: { sessionId: params?.sessionId, update: { sessionUpdate: 'session_info_update', _meta: { [EXT_META_KEY]: update } } } };
}

// The reverse of rewriteExtension: a rewritten notification arrives as session_info_update with the original update
// under EXT_META_KEY; plain session_info_updates (session title) return undefined and flow through normalize unchanged
export function extensionOf(u: acp.SessionUpdate, log?: (line: string) => void): ExtensionUpdate | undefined {
  if (u.sessionUpdate !== 'session_info_update') return undefined;
  const raw = record(record(u._meta)?.[EXT_META_KEY]);
  if (raw === undefined) return undefined;
  const kind = raw.sessionUpdate;
  if (typeof kind === 'string' && kind in ASYNC_TASK_KINDS) return asyncTaskEvent(kind, raw, log);
  if (kind === 'subagent_update' || kind === 'subagent_spawned' || kind === 'subagent_state_update') {
    const peerSessionId = str(raw.subagentSessionId);
    if (peerSessionId === undefined) {
      log?.(`${kind} without subagentSessionId dropped`);
      return { kind: 'ignored', sessionUpdate: String(kind) };
    }
    const l: SubagentLifecycle = { kind: 'lifecycle', peerSessionId, meta: raw };
    const title = str(raw.name);
    const task = str(raw.task);
    const caps = capabilitiesOf(raw.capabilities);
    const state = stateOf(raw.state, log);
    if (title !== undefined) l.title = title;
    if (task !== undefined) l.task = task;
    if (caps !== undefined) l.capabilities = caps;
    if (state !== undefined) l.state = state;
    return l;
  }
  return { kind: 'ignored', sessionUpdate: typeof kind === 'string' ? kind : '?' };
}

function asyncTaskEvent(kind: string, raw: Record<string, unknown>, log?: (line: string) => void): ExtensionUpdate {
  const asyncTaskId = str(raw.asyncTaskId);
  if (asyncTaskId === undefined) {
    log?.(`${kind} without asyncTaskId dropped`);
    return { kind: 'ignored', sessionUpdate: kind };
  }
  const e: AsyncTaskEvent = {
    kind: 'async_task', event: ASYNC_TASK_KINDS[kind]!, asyncTaskId, meta: raw,
  };
  const name = str(raw.name); if (name !== undefined) e.name = name;
  const taskType = str(raw.taskType); if (taskType !== undefined) e.taskType = taskType;
  const description = str(raw.description); if (description !== undefined) e.description = description;
  if (raw.showInTranscript === true) e.showInTranscript = true;
  if (typeof raw.canStop === 'boolean') e.canStop = raw.canStop;
  const outputFilePath = str(raw.outputFilePath); if (outputFilePath !== undefined) e.outputFilePath = outputFilePath;
  const toolCallId = str(raw.toolCallId); if (toolCallId !== undefined) e.toolCallId = toolCallId;
  const summary = str(raw.summary); if (summary !== undefined) e.summary = summary;
  const lastToolName = str(raw.lastToolName); if (lastToolName !== undefined) e.lastToolName = lastToolName;
  const usage = record(raw.usage);
  if (usage) {
    const u: NonNullable<AsyncTaskEvent['usage']> = {};
    if (typeof usage.totalTokens === 'number') u.totalTokens = usage.totalTokens;
    if (typeof usage.toolUses === 'number') u.toolUses = usage.toolUses;
    if (typeof usage.durationMs === 'number') u.durationMs = usage.durationMs;
    e.usage = u;
  }
  if (e.event === 'state') {
    const state = str(raw.state);
    if (state === undefined || !(ASYNC_TASK_STATES as readonly string[]).includes(state)) {
      log?.(`async_task_state_update with unknown state dropped: ${JSON.stringify(raw.state)}`);
      return { kind: 'ignored', sessionUpdate: kind };
    }
    e.state = state as AsyncTaskState;
  }
  return e;
}

File: rust/crates/acpira-host/src/acp/wire.rs
//! Extension session updates (mirror of src/host/acp/subagents/wire.ts). The TS host parks these kinds inside
//! `session_info_update` so the SDK's closed union does not drop them; this host has no closed union, so the raw kind is
//! read directly. The parked form is still understood (recorded fixtures carry it)

use serde_json::Value;

use acpira_shared::subagents::SubagentState;
use acpira_shared::transcript::{AsyncTaskState, TaskUsage};

pub const EXT_META_KEY: &str = "acpira/extension";

const EXT_KINDS: [&str; 6] =
  ["subagent_update", "subagent_spawned", "subagent_state_update", "async_task_spawned", "async_task_progress", "async_task_state_update"];

pub fn is_extension_kind(kind: &str) -> bool {
  EXT_KINDS.contains(&kind)
}

#[derive(Debug, Clone, PartialEq)]
pub struct SubagentLifecycle {
  pub peer_session_id: String,
  pub title: Option<String>,
  pub task: Option<String>,
  pub cancel: Option<bool>,
  pub state: Option<SubagentState>,
  pub meta: Value,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TaskEventKind {
  Spawned,
  Progress,
  State,
}

#[derive(Debug, Clone, PartialEq)]
pub struct AsyncTaskEvent {
  pub event: TaskEventKind,
  pub async_task_id: String,
  pub name: Option<String>,
  pub task_type: Option<String>,
  pub description: Option<String>,
  pub show_in_transcript: bool,
  pub can_stop: Option<bool>,
  pub output_file_path: Option<String>,
  pub tool_call_id: Option<String>,
  pub summary: Option<String>,
  pub last_tool_name: Option<String>,
  pub usage: Option<TaskUsage>,
  pub state: Option<AsyncTaskState>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExtensionUpdate {
  Lifecycle(SubagentLifecycle),
  AsyncTask(AsyncTaskEvent),
  Ignored(String),
}

fn s(v: &Value, k: &str) -> Option<String> {
  v.get(k).and_then(Value::as_str).filter(|x| !x.is_empty()).map(str::to_owned)
}

/// The raw extension update of a `session/update` payload: either the kind itself or the TS-parked form
pub fn raw_extension(update: &Value) -> Option<&Value> {
  let kind = update.get("sessionUpdate").and_then(Value::as_str)?;
  if is_extension_kind(kind) {
    return Some(update);
  }
  if kind == "session_info_update" {
    return update.get("_meta").and_then(|m| m.get(EXT_META_KEY)).filter(|r| r.is_object());
  }
  None
}

pub fn extension_of(update: &Value, log: &dyn Fn(&str)) -> Option<ExtensionUpdate> {
  let raw = raw_extension(update)?;
  let kind = raw.get("sessionUpdate").and_then(Value::as_str).unwrap_or("?");
  match kind {
    "async_task_spawned" | "async_task_progress" | "async_task_state_update" => Some(async_task_event(kind, raw, log)),
    "subagent_update" | "subagent_spawned" | "subagent_state_update" => {
      let Some(peer_session_id) = s(raw, "subagentSessionId") else {
        log(&format!("{kind} without subagentSessionId dropped"));
        return Some(ExtensionUpdate::Ignored(kind.into()));
      };
      let state = match raw.get("state") {
        None | Some(Value::Null) => None,
        Some(Value::String(v)) if matches!(v.as_str(), "running" | "completed" | "failed" | "cancelled" | "disconnected") => {
          serde_json::from_value(Value::String(v.clone())).ok()
        }
        Some(other) => {
          log(&format!("unknown subagent state {other} — treating as running"));
          Some(SubagentState::Running)
        }
      };
      let cancel = raw.get("capabilities").and_then(Value::as_object).map(|c| c.get("cancel") == Some(&Value::Bool(true)));
      Some(ExtensionUpdate::Lifecycle(SubagentLifecycle {
        peer_session_id,
        title: s(raw, "name"),
        task: s(raw, "task"),
        cancel,
        state,
        meta: raw.clone(),
      }))
    }
    other => Some(ExtensionUpdate::Ignored(other.into())),
  }
}

fn async_task_event(kind: &str, raw: &Value, log: &dyn Fn(&str)) -> ExtensionUpdate {
  let Some(async_task_id) = s(raw, "asyncTaskId") else {
    log(&format!("{kind} without asyncTaskId dropped"));
    return ExtensionUpdate::Ignored(kind.into());
  };
  let event = match kind {
    "async_task_spawned" => TaskEventKind::Spawned,
    "async_task_progress" => TaskEventKind::Progress,
    _ => TaskEventKind::State,
  };
  let usage = raw.get("usage").and_then(Value::as_object).map(|u| TaskUsage {
    total_tokens: u.get("totalTokens").and_then(Value::as_f64).map(Into::into),
    tool_uses: u.get("toolUses").and_then(Value::as_f64).map(Into::into),
    duration_ms: u.get("durationMs").and_then(Value::as_f64).map(Into::into),
  });
  let mut e = AsyncTaskEvent {
    event,
    async_task_id,
    name: s(raw, "name"),
    task_type: s(raw, "taskType"),
    description: s(raw, "description"),
    show_in_transcript: raw.get("showInTranscript") == Some(&Value::Bool(true)),
    can_stop: raw.get("canStop").and_then(Value::as_bool),
    output_file_path: s(raw, "outputFilePath"),
    tool_call_id: s(raw, "toolCallId"),
    summary: s(raw, "summary"),
    last_tool_name: s(raw, "lastToolName"),
    usage,
    state: None,
  };
  if event == TaskEventKind::State {
    let state = s(raw, "state").and_then(|v| serde_json::from_value::<AsyncTaskState>(Value::String(v)).ok());
    match state {
      Some(st) => e.state = Some(st),
      None => {
        log(&format!(
          "async_task_state_update with unknown state dropped: {}",
          raw.get("state").map(Value::to_string).unwrap_or_else(|| "undefined".into())
        ));
        return ExtensionUpdate::Ignored(kind.into());
      }
    }
  }
  ExtensionUpdate::AsyncTask(e)
}
